contracts/examples/IncentivisedSlidingWindowOracle.sol

```diff
@@ -0,0 +1,176 @@
+
+        // we only want to commit updates once per period (i.e. windowSize / granularity)
+        uint timeElapsed = block.timestamp - observation.timestamp;
+        require(timeElapsed > periodSize, 'SlidingWindowOracle: ALREADY_UPDATED_THIS_PERIOD');
+
+        (uint price0Cumulative, uint price1Cumulative,) = UniswapV2OracleLibrary.currentCumulativePrices(pair);
+        observation.timestamp = block.timestamp;
+        observation.price0Cumulative = price0Cumulative;
+        observation.price1Cumulative = price1Cumulative;
+
+        if (pair == incentivisedPair) {
+            incentiveToken.transfer(msg.sender, updateIncentiveAmount());
+        }
+
+        emit PriceUpdatedForPair(pair, tokenA, tokenB);
+    }
+
+    // returns the amount out corresponding to the amount in for a given token using the moving average over the time
+    // range [now - [windowSize, windowSize - periodSize * 2], now]
+    // update must have been called for the bucket corresponding to timestamp `now - windowSize`
+    function consult(address tokenIn, uint amountIn, address tokenOut) external view returns (uint amountOut) {
+        address pair = UniswapV2Library.pairFor(factory, tokenIn, tokenOut);
+        Observation storage firstObservation = getFirstObservationInWindow(pair);
+
+        uint timeElapsed = block.timestamp - firstObservation.timestamp;
+        require(timeElapsed <= windowSize, 'SlidingWindowOracle: MISSING_HISTORICAL_OBSERVATION'); // Not enough recorded observations
+        // should never happen.
+        require(timeElapsed >= windowSize - periodSize * 2, 'SlidingWindowOracle: UNEXPECTED_TIME_ELAPSED');
+
+        (uint price0Cumulative, uint price1Cumulative,) = UniswapV2OracleLibrary.currentCumulativePrices(pair);
+        (address token0,) = UniswapV2Library.sortTokens(tokenIn, tokenOut);
+
+        if (token0 == tokenIn) {
+            return computeAmountOut(firstObservation.price0Cumulative, price0Cumulative, timeElapsed, amountIn);
+        } else {
+            return computeAmountOut(firstObservation.price1Cumulative, price1Cumulative, timeElapsed, amountIn);
+        }
+    }
+
+    // given the cumulative prices of the start and end of a period, and the length of the period, compute the average
+    // price in terms of how much amount out is received for the amount in
+    function computeAmountOut(
+        uint priceCumulativeStart, uint priceCumulativeEnd,
+        uint timeElapsed, uint amountIn
+    ) private pure returns (uint amountOut) {
+        // overflow is desired.
+        FixedPoint.uq112x112 memory priceAverage = FixedPoint.uq112x112(
+            uint224((priceCumulativeEnd - priceCumulativeStart) / timeElapsed)
+        );
+        amountOut = priceAverage.mul(amountIn).decode144();
+    }
+
+    function getPair(address factory, address tokenA, address tokenB) public returns (address) {
+        return UniswapV2Library.pairFor(factory, tokenA, tokenB);
+    }
+}
```

Comment:
What you think of having a function to notify about new incentives added to the contract? Something in the lines of https://github.com/hermeznetwork/HEZToken/blob/main/contracts/UniPoolHEZ.sol#L125...L143